### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by N47x0**
refactored `Game` Object into interface and implementations: see `games/`

see the new `game_server/` that has a basic data model (see `DataLayer.cs`) and `Models.GameSession.cs`

Still a WIP (not integrated with frontend yet) but figured it could use some socializing at this point.

### CODE REVIEW

File: game_client/src/components/CardCollection.vue [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            v-for="(c, i) in collection"
+            :key="i"
+          >
+            <CardComp
+              :class="[owner + '-card']"
+              :card="c"
+              :isHand="true"
+              @toggle-select="onToggleSelect"
+              :highlighted="cardHighlighted(c)"
+            />
```
Comment (N47x0) [inline game_client/src/components/CardCollection.vue]: I believe this is a good opportunity for usage of `v-slot`
[vue-patterns reference](https://learn-vuejs.github.io/vue-patterns/patterns/#composition-2)

File: game_client/src/components/Hand/HandCompCards.vue [hunk 2]
```diff
@@ -1,104 +1,92 @@
...
-          </b-col>
-        </b-row>
-      </b-card>
-    </div>
+  <div class="hand-comp-cards">
+    <b-row>
+      <CardCollection 
+        :collection="getHand"
+        owner="player"
+        :highlighted="highlighted"
```
Comment (N47x0) [inline game_client/src/components/Hand/HandCompCards.vue]: think about how using slots you can avoid one layer of 'prop-drilling'

File: game_client/package.json [hunk 3]
```diff
@@ -14,6 +14,7 @@
     "axios": "^0.19.0",
     "bootstrap-vue": "^2.0.0-rc.19",
     "core-js": "^2.6.5",
+    "popper": "^1.0.1",
```
Comment (N47x0) [inline game_client/package.json:17]: what s this for?

File: game_server/Controllers/InitGameController.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        gameSession.UserPlayers.Add(ugs);
+        user_player.GameSessions.Add(ugs);
+      } 
+      // Lookup game session
+      else {
+        gameSession = _context.GameSessions.Find(sessionid);
+        initial_game_state = gameSession.GameStates.OrderBy(x => x.TurnCount).LastOrDefault();
+        // Find current/logged-in user (url param)
+        ugs = gameSession.UserPlayers.Where(x => x.User.EmailAddress == userEmail).SingleOrDefault();
+      }
```
Comment (N47x0) [inline game_server/Controllers/InitGameController.cs:80]: the intent of this entire else clause should be examined

File: game_server/Controllers/PlayGameController.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        _context = context;
+        _cardGameService = cardGame;
+    }
+
+    [EnableCors]
+    [HttpPost("GetValidPlays/{sessionid}")]
+    async public Task<ValidPlaysPayload> GetValidPlays([FromBody]ValidPlaysIncomingPayload incomingPayload){
+      Console.WriteLine(incomingPayload);
+      // TODO - Specific gamename resolution
+      var escoba_info = _context.Games.Where(x => x.GameName == "escoba").SingleOrDefault();
```
Comment (N47x0) [inline game_server/Controllers/PlayGameController.cs:38]: given `sessionid` as input the gameinfo `escoba_info` here can be retrieved from the `gameSession` once looked up

File: game_server/Models/GameSession.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    public games.GameState GameState {get; set;}
+  }
+
+  public class ValidPlaysPayload {
+    public Guid SessionId {get; set;}
+    public games.GameState GameState {get; set;}
+  }
+
+  public class ValidPlaysIncomingPayload {
+    public Guid SessionId {get; set;}
```
Comment (N47x0) [inline game_server/Models/GameSession.cs:59]: is this needed for a single field?

File: game_server/wwwroot/lib/bootstrap/dist/css/bootstrap.min.css [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
...m-select::-ms-expand{display:none}.custom-select-sm{height:calc(1.5em + .5rem + 2px);padding-top:.25rem;padding-bottom:.25rem;padding-left:.5rem;font-size:.875rem}.custom-select-lg{height:calc(1.5em + 1rem + 2px);padding-top:.5rem;padding-bottom:.5rem;padding-left:1rem;font-size:1.25rem}.custom-file{position:relative;display:inline-block;width:100%;height:calc(1.5em + .75rem + 2px);margin-bottom:0}.custom-file-input{position:relative;z-index:2;width:100%;height:calc(1.5em + .75rem + 2px);margin:0;opacity:0}.custom-file-input:focus~.custom-file-label{border-color:#80bdff;box-shadow:0 0 0 .2rem rgba(0,123,255,.25)}.custom-file-input:disabled~.custom-file-label{background-color:#e9ecef}.custom-file-input:lang(en)~.custom-file-label::after{content:"Browse"}.custom-file-input~.custom-file-label[data-browse]::after{content:attr(data-browse)}.custom-file-label{position:absolute;top:0;right:0;left:0;z-index:1;height:calc(1.5em + .75rem + 2px);padding:.375rem .75rem;font-weight:400;line-height:1.5;color:#495057;background-color:#fff;border:1px solid #ced4da;border-radius:.25rem}.custom-file-label::after{position:absolute;top:0;right:0;bottom:0;z-index:3;display:block;height:calc(1.5em + .75
```
Comment (N47x0) [inline game_server/wwwroot/lib/bootstrap/dist/css/bootstrap.min.css]: consider [this](https://docs.microsoft.com/en-us/aspnet/core/client-side/libman/libman-cli?view=aspnetcore-3.0) for removing and gitignoring wwwroot/ in the future

**Comment by shortpoet** [comment]
@N47x0 Valid plays property of game state was not being advanced within playturn.  fixed that.  then the problem was that tablecards not updating.  gettablecards.add(c) in validturn not adding c to tablecards.  tried addtablecards method that changes owner prop to table, also not working.  even just hardcoding owner to 'table' within applyplay doesn't work.  i suspect issue is with how Deck is instantiated within applyplay method.

**Pull request merged**